[RDY] Vehicle mounted washing machine

Night-Pryanik opened the pull request:
- Washing machine for now spawns only in deluxe RV. Changed it for one of the seats.
- It can hold items up to total volume of 40 liters.
- Player needs 24 charges (6 liters) of water (decided not to use clean water for washing purposes) and 5 charges of detergent. The machine is connected to the vehicle water tanks, so water is auto-drained on activation. Detergent is auto-consumed from the player's inventory.
- Washing program has a fixed duration of 1,5 hours.
- Player should wash only filthy items.
- Player can manually stop the activated machine before it's completed its program. In that case filthy items won't be cleaned, water and detergent will be simply lost.
- Player puts items in the machine and activates it. This means that he closes the lid of the machine, so he can't get his items back until washing machine is stopped - either automatically or manual.
- Game will inform player about remaining duration of the washing process every 15 minutes.
- There is issue I found - one can still put items in the activated (=closed) machine. That can be abstracted though - imagine it is the machine with upper (not side) lid that can be opened at will, even during the washing process, and player just adds some more items to the machine.
- All numbers and values are a subject for discussion.

ZhilkinSerg commented:
Gorgon says that something is not good with astyle:

````
21:09:02 astyle regressions found.
21:09:02 Formatted  src/pickup.cpp
21:09:02 Makefile:950: recipe for target 'astyle-check' failed
21:09:02 make: *** [astyle-check] Error 1
````

AMurkin commented:
Also it ironically states `21:09:00 +./mainline-build.zsh:87> make clean`

ZhilkinSerg commented on src/map.cpp:
```diff
@@ -4672,6 +4672,32 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
+            cur_veh->parts[part].enabled = false;
+            counter = 0;
+        } else if( calendar::once_every( MINUTES( 15 ) ) ) {
+            add_msg( _("It should take %d minutes to finish washing."), time_left / MINUTES( 1 ) + 1 );
```
Whitespaces are probably required in `_` function?

Night-Pryanik replied:
Fixed.

Coolthulhu commented:
Statics are unacceptable here.
Consider the case where two vehicles have washing machines:
* Turn on first machine
* Turn on second machine
* First machine increments static counter
* Second machine increments same exact static counter
* Increments repeat for `washing_time / 2`
* First washing machine finishes washing in `washing_time / 2`
* Second machine finishes washing in `washing_time * 3 / 2`

This only gets worse as more machines are in the bubble.

The easiest way to fix it would be to have washing machine note in its item (`vehicle_part::base`) when washing started and use this number.

Night-Pryanik commented:
> The easiest way to fix it would be to have washing machine note in its item (vehicle_part::base) when washing started and use this number.

Umm, need some help on this.

Coolthulhu commented on src/pickup.cpp:
```diff
@@ -83,6 +83,9 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
     const bool is_convertible = ( veh->tags.count( "convertible" ) > 0 );
     const bool remotely_controlled = g->remoteveh() == veh;
+    const bool has_washmachine = ( veh->part_with_feature( veh_root_part, "WASHING_MACHINE" ) >= 0 );
+    bool working = false;
```
`washing_machine_on` or something like that. Ambiguous variable names make code harder to read.

Night-Pryanik replied:
Well, I'm not against renaming, but the name `has_washmachine` was taken by an analogy of `has_kitchen`, `has_weldrig` and so on just above.

AMurkin replied:
I think it's about `working` flag

Night-Pryanik replied:
Oh, sorry, mislooked.

Coolthulhu commented on src/pickup.cpp:
```diff
@@ -83,6 +83,9 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
     const bool remotely_controlled = g->remoteveh() == veh;
+    const bool has_washmachine = ( veh->part_with_feature( veh_root_part, "WASHING_MACHINE" ) >= 0 );
+    bool working = false;
+    bool detergent_is_enough = false;
```
Move this variable into the case.
You can have variable declarations inside a case if you wrap the block in braces, like this:
```
switch( x ) {
    case y: {
        bool is_happening = true;
        if( is_happening ) add_msg( m_good, "It's happening!" );
    }
}
```

Night-Pryanik replied:
> Move this variable into the case.

You mean move `bool detergent_is_enough = false` into the `USE_WASHMACHINE` case?

Coolthulhu replied:
Yes.
Nothing else uses it.

Night-Pryanik replied:
Moving `bool detergent_is_enough = false` into the `USE_WASHMACHINE case` generates `crosses initialization of 'bool detergent_is_enough'` error.

Coolthulhu replied:
As I said: you need an extra pair of braces inside the case.
Look at my example above.

Night-Pryanik replied:
Ok, thanks. Fixed in 3af7934.

Coolthulhu commented on src/pickup.cpp:
```diff
@@ -169,6 +179,34 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+                    veh->tags.erase( "manual_stop" );
+                    add_msg( m_neutral, _( "You close the lid of the washing machine, and turn it on.  The washing machine is being filled with soapy water." ) );
+                }
+            }
```
Would be better if you moved the whole block to `interact_with_vehicle` in `pickup.cpp`.
Washing machine should not require you to go to vehicle controls to activate it, but rather should be interacted with directly.

Night-Pryanik replied:
But it is already in `interact_with_vehicle` in `pickup.cpp`. And it indeed working by interacting with it directly, not through controls.

Coolthulhu replied:
Oh, my bad then

Coolthulhu commented:
> Umm, need some help on this.

This isn't exactly easy. Vehicle parts aren't easy to work with.

I think I have an easier idea, though:
* In `interact_with_vehicle` in `pickup.cpp`, check if there is a washing machine
* Check if the washing machine is `enabled`
* If washing machine is not enabled, it means it isn't working (duh). Proceed with all the checks for water and detergent. If everything is OK, demand that the player removes all non-filthy items (if any). If there are none, ask the player if they want to enable it. If player wants to enable it, close and enable the machine, and set birthday of all items (or just the first item) inside machine to current turn
* If the machine is enabled, it is working. Player can open it, but the items inside will still be dirty AND the machine will become disabled.
* In `vehicle::idle`, check if the first item in machine is already washed. Like this: `first_item.bday + washing_time > current_turn`. If it is, disable the machine and clean all the items in machine.

That way it would work similarly to furniture charcoal kiln.

Night-Pryanik commented:
I was able to get rid of the static counters.
Activating the machine set the bday of the filthy item to the current day. Then in `process_vehicle_items` the game checks for the `time_left == washing_time - calendar::turn.get_turn() + bday`. If there is no time left, stop the process, and clean the filthy clothes.

Coolthulhu commented:
There are still some issues:
* You still have to interact with vehicle controls, which is not how you turn on a washing machine IRL. Moving the thing to pickup would fix that. This is not just a realism issue: it's an UI issue that adds an extra unnecessary step.
* You can wash - and thus apply bday change - to all items, including those that rely on bday to do calculations (food, corpses). You could wash a stack of squirrel corpses to freshen them.
* "Washing machine has finished" x20, because you apply finishing per item. Either check only the first item or check them all, then clean them up in a separate loop, then print one message.
* Which machine has finished washing? When batteries die, they say which vehicle lost battery power.

Night-Pryanik commented:
> You can wash - and thus apply bday change - to all items, including those that rely on bday to do calculations (food, corpses). You could wash a stack of squirrel corpses to freshen them.

Fixed in abaca20.

> "Washing machine has finished" x20, because you apply finishing per item. Either check only the first item or check them all, then clean them up in a separate loop, then print one message.

Fixed in da82944.

Night-Pryanik commented:
I just can't to figure out what are the problems with astyle in `pickup.cpp`.

Coolthulhu commented:
It would be easiest if you just downloaded the astyle binary version 2.05.1 and ran it (from command line), like this:
```
astyle.exe --options=path\to\Cataclysm-DDA\.astylerc -n path\to\Cataclysm-DDA\src\pickup.cpp
```

Astyle sometimes has really stupid requirements that you won't figure out yourself. For example, it may require very strict line breaks that look ugly, or weird alignment.

Coolthulhu commented on data/json/vehicle_parts.json:
```diff
@@ -694,6 +694,28 @@
+            {"item": "steel_chunk", "count": [8, 13]},
+            {"item": "scrap", "count": [8, 13]},
+            {"item": "hose", "prob": 50}
+        ]
```
Not necessary in this PR, but it's a good idea to add disassemble recipes to new things, so that they always have some use, even the duplicates.

Coolthulhu commented on src/map.cpp:
```diff
@@ -4672,6 +4672,28 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
+                washing_machine_finished = true;
+                cur_veh->parts[part].enabled = false;
+            } else if( calendar::once_every( MINUTES( 15 ) ) ) {
+                add_msg( _("It should take %d minutes to finish washing."), time_left / MINUTES( 1 ) + 1 );
```
This one could use the vehicle name too.
Consider a base with 3 washing machines on wheels. It's unlikely to happen (that detergent cost isn't low), but it's better to inform the player too much than too little, as long as it doesn't require more messages.

Night-Pryanik replied:
Fixed in 4ab151e.

AMurkin commented on src/map.cpp:
```diff
@@ -4672,6 +4672,28 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
+            }
+        }
+        if( washing_machine_finished ) {
+            add_msg( _( "The washing machine in the %s has finished washing." ), cur_veh->name.c_str() );
```
You can move loop for items here. And check only first item above.
```
for( auto &n : cur_veh->get_items( part ) )
  n.item_tags.erase( filthy );
```

Night-Pryanik replied:
> And check only first item above.
Wanna know how to do this.

Coolthulhu replied:
```
auto &items = cur_veh->get_items( part );
if( items.empty() || items.front().bday + time > turn ) {
    for( item & cur_item : items ) {
        cur_item.item_tags.erase( filthy );
    }
    add_msg( "done" );
    turn_off_machine();
}

Coolthulhu replied:
Though that would require the first item to be filthy and have bday set, meaning you'd either have to sort it to the front and require that at least one such item exists or refuse to allow washing ANY non-filthy items.

It would be much faster to process, though.

AMurkin replied:
I think
`item n = cur_veh->get_items( part ).front();`

AMurkin commented on src/pickup.cpp:
```diff
@@ -169,6 +181,50 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+            bool detergent_is_enough = false;
+            bool filthy_items = false;
+            if( g->u.crafting_inventory().has_charges( "detergent", 5 ) ) {
+                detergent_is_enough = true;
```
Is that too long?
`bool detergent_is_enough = g->u.crafting_inventory().has_charges( "detergent", 5 );`

Night-Pryanik replied:
Implemented in f612ae5.

Night-Pryanik commented:
Will my current implementation suffice?
Should I add comments to my added code?

Coolthulhu commented:
Looks good enough on paper.
If it doesn't have bugs, it should be OK.

BevapDin commented on src/pickup.cpp:
```diff
@@ -95,7 +98,16 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
 
-    if( from_vehicle ) {
+    for( auto e : veh->get_parts( "WASHING_MACHINE" ) ) {
+        washing_machine_on = e->enabled;
```
This won't work correctly if there are more than one washing machines installed. It will use the `enabled` status of the last one, even if you examine a different one.

Night-Pryanik replied:
Any suggestions?

Coolthulhu replied:
Save index of the washmachine here.
You get it with `veh->part_with_feature( veh_root_part, "WASHING_MACHINE" )` above, but then you drop it after comparison.

Night-Pryanik replied:
Damn, still can't understand what I need to do here.

Night-Pryanik replied:
I think I fixed it in d5c7d62.

codemime commented on src/map.cpp:
```diff
@@ -4672,6 +4672,28 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
+            const int washing_time = MINUTES( 90 );
+            int time_left = washing_time - calendar::turn.get_turn() + n.bday;
+            static const std::string filthy( "FILTHY" );
+            if( time_left == 0 ) {
```
This condition would probably "break" the washing machine if one leaves the reality bubble and returns more than 90 minutes later. `time_left` then becomes negative and provokes infinite washing cycle (along with increasingly negative number of minutes in the message).

Night-Pryanik replied:
Will changing it to `time_left <= 0` prevent this?

codemime replied:
I guess so.

codemime replied:
BTW `time_left` can be made `const` too.

Night-Pryanik replied:
> BTW time_left can be made const too.
But it changes every turn?

codemime replied:
Yes, but it's scoped, non-static, and it doesn't change between `{` and `}`, so it can/should be made `const`.

Night-Pryanik replied:
Done in b33a11c and 42c9538.

codemime commented on src/pickup.cpp:
```diff
@@ -169,6 +181,44 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+            for( auto &i : veh->get_items( veh->part_with_feature( veh_root_part, "WASHING_MACHINE" ) ) ) {
+                static const std::string filthy( "FILTHY" );
+                if( i.has_flag( filthy ) ) {
+                    i.bday = calendar::turn.get_turn();
```
Washing may be cancelled because of insufficient amount of water/detergent. Why reset `bday` in this case?

Night-Pryanik replied:
Because items in the washing machine (`i`) are only accessible inside `for` cycle above, but it needed to be set below, in the `else if( !filthy_items )` check.

Coolthulhu replied:
Split it into two loops - one for check, one for bday setting.
Also save a reference to the result of `get_items`, don't re-get it every time.

codemime replied:
I don't see a contradiction here. There's plenty of ways to check before and set after (two `for` loops is one of them). Even better `std::any_of` to check and a `for` to set.

Night-Pryanik replied:
Fixed in 2d1e8b1.

codemime commented on src/pickup.cpp:
```diff
@@ -169,6 +181,44 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+                    add_msg( m_bad, _( "You need 5 charges of detergent for the washing machine." ) );
+                } else if( !filthy_items ) {
+                    add_msg( m_bad,
+                             _( "There are only non-filthy items in the washing machine.  There is no need to wash them." ) );
```
This check looks overzealous. IMHO players should be allowed to do even nonsensical things, including washing clean clothes.

Night-Pryanik replied:
@Coolthulhu says that player must wash only filthy items, you are telling right the opposite. I don't know what to prefer :-)

codemime replied:
That was an IMHO. My opinion on the matter is not paramount, but I do stick to it. It's a roguelike after all.

Coolthulhu replied:
The implementation relies on existence of at least one filthy item.
The filthy item serves as turn counter.

codemime replied:
If that's the only reason, perhaps a dummy item would do the trick.

codemime commented on src/map.cpp:
```diff
@@ -4672,6 +4672,28 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
+                washing_machine_finished = true;
+                cur_veh->parts[part].enabled = false;
+            } else if( calendar::once_every( MINUTES( 15 ) ) ) {
+                add_msg( _("It should take %d minutes to finish washing items in the %s."), time_left / MINUTES( 1 ) + 1, cur_veh->name.c_str() );
```
I'd put this message in the title of `selectmenu` where it could be seen when examining the washing machine tile.

codemime commented on src/pickup.cpp:
```diff
@@ -169,6 +181,44 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
             return DONE;
 
+        case USE_WASHMACHINE: {
+            bool detergent_is_enough = g->u.crafting_inventory().has_charges( "detergent", 5 );
```
Speaking of code, I'd turn the whole snippet into a named function to prevent further bloating of `interact_with_vehicle`.

Night-Pryanik replied:
Fixed in 8cc9d3f.

Coolthulhu commented on src/pickup.cpp:
```diff
@@ -169,6 +181,44 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+        case USE_WASHMACHINE: {
+            bool detergent_is_enough = g->u.crafting_inventory().has_charges( "detergent", 5 );
+            bool filthy_items = false;
+            for( auto &i : veh->get_items( veh->part_with_feature( veh_root_part, "WASHING_MACHINE" ) ) ) {
```
Don't re-get `part_with_feature`. Save it once and reuse that.

Night-Pryanik replied:
Fixed in 2d1e8b1.

Coolthulhu commented on src/pickup.cpp:
```diff
@@ -169,6 +181,44 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+                    add_msg( m_bad,
+                             _( "You turn the washing machine off before it's finished the program, and open its lid." ) );
+                } else if( veh->fuel_left( "water" ) < 24 ) {
+                    add_msg( m_bad, _( "You need 24 charges of water to fill the washing machine." ) );
```
Needs to specifically mention that the water is supposed to be in vehicle tanks.
Could also allow clean water, as there is no good way of dirtying clean water.

Night-Pryanik replied:
Fixed in 56d81d9.

Coolthulhu commented on src/map.cpp:
```diff
@@ -4672,6 +4672,28 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
     }
+
+    const bool washmachine_here = cur_veh->part_flag( part, VPFLAG_WASHING_MACHINE ) && cur_veh->has_part( "WASHING_MACHINE", true );
+    bool washing_machine_finished = false;
```
You don't use it outside the `if`, so it should be inside the `if`.

Night-Pryanik replied:
You mean outside of `if( time_left <= 0 )` or `if( finished )`?

Coolthulhu commented on src/map.cpp:
```diff
@@ -4672,6 +4672,28 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
         }
     }
+
+    const bool washmachine_here = cur_veh->part_flag( part, VPFLAG_WASHING_MACHINE ) && cur_veh->has_part( "WASHING_MACHINE", true );
```
This check is not correct.
Consider more than one washing machine: in this case this check is `true` if any of the washing machines is turned on.

The check should be:
```
if( cur_veh->part_flag( part, VPFLAG_WASHING_MACHINE ) && cur_veh->is_part_on( part ) ) {
```

Night-Pryanik replied:
Fixed in fb0146f.

codemime commented on src/pickup.cpp:
```diff
@@ -169,6 +182,49 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+            bool detergent_is_enough = g->u.crafting_inventory().has_charges( "detergent", 5 );
+            static const std::string filthy( "FILTHY" );
+            bool filthy_items = std::any_of( veh->get_items( washing_machine_part ).begin(),
+                                             veh->get_items( washing_machine_part ).end(),
```
Though the function doesn't look expensive, it would be safer to stash the result of `veh->get_items( washing_machine_part ) ` in a temp variable: `const auto items = veh->get_items( washing_machine_part )`.

Night-Pryanik replied:
Implemented in 6114c40, but had to made it without `const`, as `items` are changed below, where we set the `bday`.

AMurkin commented on src/pickup.cpp:
```diff
@@ -169,6 +179,46 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+            } else {
+                veh->parts[washing_machine_part].enabled = true;
+                for( auto &n : items ) {
+                    if( filthy_items ) {
```
This condition is redundant.

AMurkin commented on src/pickup.cpp:
```diff
@@ -169,6 +179,46 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
+                add_msg( m_bad, _( "You need 5 charges of detergent for the washing machine." ) );
+            } else if( !filthy_items ) {
+                add_msg( m_bad,
+                         _( "There are only non-filthy items in the washing machine.  There is no need to wash them." ) );
```
Now that message is incorrect.

Night-Pryanik replied:
Fixed in 1163892.

Night-Pryanik commented:
Jenkins, rebuild.

cainiaowu commented:
Jenkins, rebuild.

codemime commented on src/map.cpp:
```diff
@@ -4677,15 +4677,15 @@ static void process_vehicle_items( vehicle *cur_veh, int part )
         for( auto &n : cur_veh->get_items( part ) ) {
-            const int washing_time = MINUTES( 90 );
-            const int time_left = washing_time - n.age();
+            const time_duration washing_time = 2_hours;
```
`2_hours` isn't equivalent to `MINUTES( 90 )`. Is that intentional?

Night-Pryanik replied:
I don't know how to translate 90 minutes in to new for.

ZhilkinSerg replied:
> don't know how to translate 90 minutes in to new for.

`90_minutes` and most probably `1.5_hours` will also work

AMurkin replied:
90_minutes is ok. _hours take int so 1.5_hours is not ok.

Night-Pryanik commented:
I'm sorry for possibly being annoying, but is there something critical that blocks this from merging?

ZhilkinSerg commented:
> I'm sorry for possibly being annoying, but is there something critical that blocks this from merging?

Possibly lack of time for contributors with merge rights.

kevingranade commented on src/pickup.cpp:
```diff
@@ -83,9 +83,13 @@ interact_results interact_with_vehicle( vehicle *veh, const tripoint &pos,
     const bool remotely_controlled = g->remoteveh() == veh;
+    const int washing_machine_part = veh->part_with_feature( veh_root_part, "WASHING_MACHINE" );
+    const bool has_washmachine = washing_machine_part >= 0;
+    bool washing_machine_on = veh->parts[washing_machine_part].enabled;
```
Segfault on this line, you cannot dereference an array with the -1 returned by veh->part_with_feature( veh_root_part, "WASHING_MACHINE" ); if there is no washing machine in the vehicle.
This would work:
```
bool washing_machine_on = ( washing_machine_part == -1 ) ? false :  veh->parts[washing_machine_part].enabled;
```

kevingranade replied:
I fixed this during the merge, looks like it works flawlessly otherwise :+1:

Night-Pryanik replied:
Thanks!
Though I'd like to know at which moment does it segfault?
